src/Transition.js

```diff
@@ -361,7 +361,11 @@ class Transition extends React.Component {
     delete childProps.onExited
 
     if (typeof children === 'function') {
-      return children(status, childProps)
+      return children(
+        status,
+        childProps,
+        () => this.setState({status: ENTERED})
```

Comment:
this is too simple unfortunately. The done callback really needs to be the same as the actual done callback, otherwise state won't get cleaned up correct. You also need to handle the exiting case as well. I would start by doing something like how setNextCallback works in `onTransitionEnd` and then passing it here from the instance, making sure it gets nulled out when it's not in the entering or exiting state

Reply:
> relates to race conditions when triggering onEnter*/onExit* hooks

yes and you need to avoid race conditions here as well, you shouldn't be able to call this twice, or after the animation completes